Added introduce action

7h3kk1d opened the pull request:
### Introduce: Fill Empty Holes with Canonical Forms for Expressions and patterns

This PR adds Introduce, a feature that fills an empty hole expression under the cursor with a canonical form based on its expected type. This helps scaffold expressions quickly without manually constructing them.

Supported Expressions:

-  Arrow types `T1 -> T2 `:  `fun ? -> ?`
-  Tuple types `(T1, T2, ..., Tn)`:  `(? , ? , ... , ?)`
    -  Note: Labels are not inserted, as they are inferred and it's unclear whether they should be included in the syntax.
- Singleton variants: Replaced with the corresponding constructor
  - Nullary constructor `+A`: `A`
  - Non-nullary constructor `+A(T)`: `A(?)`
- Forall types `forall ? -> ?` → Replaced with typfun ? -> ?

The Introduce command is available in the command palette and is bound to Ctrl/Cmd + I for quick access.

[Screencast From 2025-03-16 13-34-59.webm](https://github.com/user-attachments/assets/795942e6-70ca-4a9e-8ce1-0c822aeeb627)



### Future Work

- ~Support for patterns~ Bassic patterns were added
- Add a Destruct (enumerate cases) action
- Consider adding a recursive introduce to scaffold deep structures

7h3kk1d committed:
Added refine for products and arrows

7h3kk1d commented on src/haz3lcore/zipper/action/Perform.re:
```diff
@@ -30,6 +30,32 @@ let go_z =
+          switch (Typ.term_of(Typ.weak_head_normalize(ctx, ana))) {
+          | Arrow(_, _) => "fun -> "
+          | Prod(tys) =>
+            "(" ++ StringUtil.repeat(List.length(tys) - 1, ", ") ++ ") "
```
This ends up adding a trailing space but that also seems to happen in the paste case as well.

7h3kk1d replied:
We don't want extra parens if we're inside of a function application but I don't know how much it matters.

cyrus- commented:
- let's call it Introduce instead of Refine
- could also support single-constructor sum types

cyrus- commented:
- also add typfun for polymorphic functions

cyrus- commented:
- don't add extra parens in ap position

7h3kk1d commented:
Add case patterns

7h3kk1d committed:
Merge branch 'dev' into refine

7h3kk1d committed:
Merge branch 'dev' into refine

7h3kk1d committed:
Merge branch 'dev' into refine

7h3kk1d committed:
Merge branch 'dev' into refine

7h3kk1d committed:
Rename 'Refine' action to 'Introduce'

7h3kk1d committed:
Move introduce code to its own module

7h3kk1d committed:
Add tests for introduce_expression

7h3kk1d committed:
Add Test_Introduce to test suite

7h3kk1d committed:
Beginning zipper tests of introduce

7h3kk1d committed:
Finish first introduce test

7h3kk1d committed:
Add test case for function introduction

7h3kk1d committed:
Enhance introduce functionality to handle already parenthesized expressions

7h3kk1d committed:
Change tests to assume current term is selected

7h3kk1d committed:
introduce action to select current term before introducing

7h3kk1d committed:
Add support for singleton variants in introduce_expression and corresponding test case

7h3kk1d committed:
Add singleton variant with payload to introduce

7h3kk1d committed:
Fix test for singleton variant

7h3kk1d committed:
Add hotkey for Introduce shortcut

7h3kk1d committed:
remove variable

7h3kk1d committed:
Add forall typfuns to introduce action

7h3kk1d committed:
Use FreshGrammar in Introduce.introduce_expression

7h3kk1d committed:
Refactor introduce function to use let+

7h3kk1d committed:
Cleanup test

7h3kk1d committed:
Merge remote-tracking branch 'origin/dev' into refine

7h3kk1d committed:
Merge commit 'f478f8d3dc371a37969dfdc8010018df1b5185ef' into refine

7h3kk1d committed:
Run reformat

7h3kk1d committed:
Merge remote-tracking branch 'origin/dev' into refine

7h3kk1d commented on src/haz3lweb/app/input/Shortcut.re:
```diff
@@ -251,6 +251,13 @@ let shortcuts = (sys: Util.Key.sys): list(t) =>
+    mk_shortcut(
+      ~mdIcon="bolt",
+      ~section="Refactoring",
+      ~hotkey=Keyboard.meta(sys) ++ "+i",
```
This is still doing ctrl on my mac

7h3kk1d replied:
Fixed in https://github.com/hazelgrove/hazel/pull/1416/commits/0eabe893237b64e29a108ac0c5691c277c79c496. Was broken for all shortcuts.

7h3kk1d committed:
Reorder initialization in on_startup so os is set before ninjakeys is initialized

7h3kk1d commented on src/haz3lweb/Main.re:
```diff
@@ -138,13 +138,13 @@ let start = {
-    NinjaKeys.initialize(Shortcut.options(schedule_action));
-    JsUtil.focus_clipboard_shim();
     Os.is_mac :=
       Dom_html.window##.navigator##.platform##toUpperCase##indexOf(
         Js.string("MAC"),
       )
       >= 0;
+    NinjaKeys.initialize(Shortcut.options(schedule_action));
+    JsUtil.focus_clipboard_shim();
```
@disconcision any concern with focusing clipboard after setting OS?

7h3kk1d committed:
Merge branch 'dev' into refine

7h3kk1d committed:
Merge branch 'dev' into refine

7h3kk1d commented:
Add labels to tuples. Add toggle lablels action. Set cursor to first hole generated.

7h3kk1d commented:
Somehow deleting expression on f( when f has type (Int, Int) -> Int

7h3kk1d committed:
Prevent replacing non-empty hole expressions in introduce function

7h3kk1d committed:
Add support for empty list and string literals in introduce_expression

7h3kk1d committed:
Merge branch 'dev' into refine

7h3kk1d commented:
> Somehow deleting expression on f( when f has type (Int, Int) -> Int

I have resolve this issue by blocking the introduce action when the cursor would select something that isn't a hole. The issue was that if you do a select term on a function application with no right paren the cursor inspector is left on the hole even though the selection isn't a hole:

![Screenshot From 2025-03-28 10-55-55](https://github.com/user-attachments/assets/5184a1bd-e993-4576-abd4-1c15f9383412)

It's possible that we should be selecting an adjacent hole rather than just the current expression under the cursor but I don't want to break expectations with the cursor inspector yet.

> Add labels to tuples, add empty string, empty list

These are done.

7h3kk1d committed:
Add labels to labeled tuples in introduce

7h3kk1d committed:
Introduce action now places the cursor on a specific expression

7h3kk1d committed:
Add documentation for introduce_expression function

7h3kk1d committed:
[Needs Discussion] Introduce now places the cursor inside of empty lists and empty strings

- This moves code from inside of Move.Make to top-level in Move
- The code doesn't rely on `Measured`, `TermRanges` so it's unnecessary to compute those for the functions. I need to discuss with Andrew or whoever cares about how this should be laid out.

7h3kk1d commented on src/haz3lcore/zipper/action/Move.re:
@dm0n3y @disconcision I would love your thoughts on how I should be handling this. I'm using `Move.primary` in `Introduce` and it doesn't have any dependency on the `Measured` or `TermRanges` coming in from `Move.S`. So I just pulled all of those functions out of the `Make` functor. Alternatively some of these can be pushed into the `Zipper` module but overall it's annoying to have the required dependencies when the functions don't require them.

7h3kk1d committed:
Remove test case for partially parenthesized function

7h3kk1d committed:
Remove debug print statement from introduce function

7h3kk1d committed:
Remove comment

7h3kk1d committed:
ExpToSegment to handle unit patterns and already parenthesized patterns

7h3kk1d committed:
Some patterns get generated on introduce action

7h3kk1d committed:
Cleaning up duplication

7h3kk1d committed:
Pull introduce logic into a functor

7h3kk1d committed:
Cleanup

7h3kk1d committed:
Merge remote-tracking branch 'origin/dev' into refine

7h3kk1d commented on src/haz3lcore/pretty/ExpToSegment.re:
```diff
@@ -397,7 +397,8 @@ and parenthesize_pat =
   | EmptyHole
-  | Constructor(_) => pat
+  | Constructor(_)
+  | Tuple([]) => pat
```
This has unit patterns rendered as () as opposed to (())

7h3kk1d committed:
Merge branch 'dev' into refine
